pkg/kubelet/dockershim/docker_container_windows.go

```diff
@@ -0,0 +1,180 @@
+	if err != nil {
+		return err
+	}
+
+	if createConfig.HostConfig == nil {
+		createConfig.HostConfig = &dockercontainer.HostConfig{}
+	}
+
+	createConfig.HostConfig.SecurityOpt = append(createConfig.HostConfig.SecurityOpt, "credentialspec=registry://"+valueName)
+	cleanupInfo.gMSARegistryValueName = valueName
+
+	return nil
+}
+
+const (
+	registryNamePrefix = "k8s-cred-spec-"
+	// same as https://github.com/moby/moby/blob/93d994e29c9cc8d81f1b0477e28d705fa7e2cd72/daemon/oci_windows.go#L23
+	credentialSpecRegistryLocation = `SOFTWARE\Microsoft\Windows NT\CurrentVersion\Virtualization\Containers\CredentialSpecs`
+)
+
+// useful to allow mocking the registry in tests
+type registryKey interface {
+	SetStringValue(name, value string) error
+	DeleteValue(name string) error
+	Close() error
+}
+
+var registryCreateKeyFunc = func(baseKey registry.Key, path string, access uint32) (registryKey, bool, error) {
+	return registry.CreateKey(baseKey, path, access)
+}
+
+// and same for random
+var randomReader = rand.Reader
+
+// copyGMSACredSpecToRegistryKey copies the credential specs to a unique registry value, and returns its name.
+// To avoid leaking registry keys over the life of the node, we generate a unique name for that value, and clean
+// it up after creating the container.
+func copyGMSACredSpecToRegistryValue(credSpec string, dockerContainerName string) (string, error) {
+	valueName, err := gMSARegistryValueName(credSpec, dockerContainerName)
+	if err != nil {
+		return "", err
+	}
+
+	// write to the registry
+	key, _, err := registryCreateKeyFunc(registry.LOCAL_MACHINE, credentialSpecRegistryLocation, registry.SET_VALUE)
+	if err != nil {
+		return "", fmt.Errorf("unable to open registry key %s: %v", credentialSpecRegistryLocation, err)
+	}
+	defer key.Close()
+	if err = key.SetStringValue(valueName, credSpec); err != nil {
+		return "", fmt.Errorf("unable to write into registry value %s/%s: %v", credentialSpecRegistryLocation, valueName, err)
+	}
+
+	return valueName, nil
+}
+
+// gMSARegistryValueName computes the name of the registry value where to store the GMSA cred spec contents.
+// The value's name is computed by concatenating the docker container's name (guaranteed to be unique over the
+// container's lifetime), the value itself, and an additional 64 random bytes.
+func gMSARegistryValueName(inputs ...string) (string, error) {
+	hasher := sha256.New()
+	for _, s := range inputs {
+		// according to the doc, that can never return an error
+		io.WriteString(hasher, s)
+	}
+	randBytes := make([]byte, 64)
+	if _, err := randomReader.Read(randBytes); err != nil {
+		return "", fmt.Errorf("unable to generate random string: %v", err)
+	}
+	hasher.Write(randBytes)
+
+	return registryNamePrefix + hex.EncodeToString(hasher.Sum(nil)), nil
+}
+
+// performPlatformSpecificContainerCreationCleanup is responsible for doing any platform-specific cleanup
```

Comment:
I think a wrapper around `performPlatformSpecificContainerCreationCleanup` that goes and cleans up all credspec entries (created specifically by kubelet) needs to be invoked during kubelet initialization on the host - somewhere around: https://github.com/kubernetes/kubernetes/blob/f0beaf46db82e713392a5c0f880e27a0e14c0690/pkg/kubelet/kubelet.go#L619 This will handle cases where the kubelet may have crashed around container creation right after credspecs got persisted.

Reply:
👍